### PR TITLE
CTCP-4669: Bug 2

### PR DESCRIPTION
**Pull request by joecefnhedog**
Fix for:

> Bug 2
> Scenario - the Active Border Means information is provided via the xml
> When you select [Change] Identification number it sends you to an automatic sign out

### CODE REVIEW

**Comment by joecefnhedog** [comment]
Screenshot 2024-01-03 at 14 50 12

**Comment by joecefnhedog** [comment]
```
<ActiveBorderTransportMeans>
    <identificationNumber>TestIdent</identificationNumber>
</ActiveBorderTransportMeans>

```

File: app/controllers/transport/border/active/IdentificationNumberController.scala [hunk 1]
```diff
@@ -87,4 +105,32 @@ class IdentificationNumberController @Inject() (
...
+
+  private def identificationPageIe015(activeIndex: Index)(implicit request: MandatoryDataRequest[_]): Option[String] =
+    request.userAnswers.departureData.Consignment.ActiveBorderTransportMeans.flatMap(
+      x => x.lift(activeIndex.position).flatMap(_.typeOfIdentification)
+    )
+
+  private def getReferenceDataFor15(activeIndex: Index)(implicit request: MandatoryDataRequest[_]): Option[Future[Identification]] =
+    identificationPageIe015(activeIndex).map {
+      aaa =>
+        service.getBorderMeansIdentification(aaa)(hc(request.request))
```
Comment (jackheslop96) [inline app/controllers/transport/border/active/IdentificationNumberController.scala:119]: Do you need to be using `MandatoryDataRequest`? If you change these to `DataRequest` you can probably get rid of `(hc(request.request))`

File: app/controllers/transport/border/active/IdentificationNumberController.scala [hunk 2]
```diff
@@ -87,4 +105,32 @@ class IdentificationNumberController @Inject() (
...
+    request.userAnswers.get(IdentificationPage(activeIndex))
+
+  private def identificationPageIe015(activeIndex: Index)(implicit request: MandatoryDataRequest[_]): Option[String] =
+    request.userAnswers.departureData.Consignment.ActiveBorderTransportMeans.flatMap(
+      x => x.lift(activeIndex.position).flatMap(_.typeOfIdentification)
+    )
+
+  private def getReferenceDataFor15(activeIndex: Index)(implicit request: MandatoryDataRequest[_]): Option[Future[Identification]] =
+    identificationPageIe015(activeIndex).map {
+      aaa =>
```
Comment (jackheslop96) [inline app/controllers/transport/border/active/IdentificationNumberController.scala:118]: Can this be named better?

File: app/services/MeansOfTransportIdentificationTypesActiveService.scala [hunk 3]
```diff
@@ -33,6 +34,14 @@ class MeansOfTransportIdentificationTypesActiveService @Inject() (referenceDataC
   ): Future[Seq[Identification]] =
     referenceDataConnector.getMeansOfTransportIdentificationTypesActive().map(filter(_, index, borderModeOfTransport)).map(sort)
 
+  def getBorderMeansIdentification(code: String)(implicit hc: HeaderCarrier): Future[Identification] =
+    for {
+      qualifiers <- referenceDataConnector.getMeansOfTransportIdentificationTypesActive()
+      qualifierFound = qualifiers
+        .find(_.code == code)
+        .getOrElse(referenceDataException("borderMeansIdentification", code, qualifiers))
```
Comment (jackheslop96) [inline app/services/MeansOfTransportIdentificationTypesActiveService.scala:42]: Could we instead add a method to the connector to user `/filtered-lists` for the `TypeOfIdentificationofMeansOfTransportActive` code list? The filter gets passed through all the way to the mongo query - more efficient than handling the entire list in memory.

The method would still return a Seq[Identification]. Perhaps we could refactor to return NonEmptyList / NonEmptySeq? Then in the service method we can safely call `.head` without having to do a `.getOrElse`

Comment (jackheslop96) [reply app/services/MeansOfTransportIdentificationTypesActiveService.scala:42]:
```
  implicit def responseHandlerGeneric[A](implicit reads: Reads[A]): HttpReads[NonEmptySeq[A]] =
    (_: String, _: String, response: HttpResponse) => {
      response.status match {
        case OK =>
          (response.json \ "data").validate[Seq[A]] match {
            case JsSuccess(Nil, _) =>
              throw new NoReferenceDataFoundException
            case JsSuccess(head :: tail, _) =>
              NonEmptySeq(head, tail)
            case JsError(errors) =>
              throw JsResultException(errors)
          }
        case e =>
          logger.warn(s"[ReferenceDataConnector][responseHandlerGeneric] Reference data call returned $e")
          throw new Exception(s"[ReferenceDataConnector][responseHandlerGeneric] $e - ${response.body}")
      }
    }
```

File: app/services/ReferenceDataNotFoundException.scala [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package services
+
+class ReferenceDataNotFoundException[T](refName: String, refDataCode: String, listRefData: Seq[T])
```
Comment (jackheslop96) [inline app/services/ReferenceDataNotFoundException.scala:19]: See above comment. This class may become obsolete.

**Review by jackheslop96 — APPROVED** [review]

**Pull request merged**